### PR TITLE
Add OP Stack contracts to ChainConstants.contracts

### PR DESCRIPTION
**Pull request by wilsoncusack**
This will unlock improved APIs and typing in [op-viem](https://github.com/base-org/op-viem).

---

## PR-Codex overview
This PR focuses on adding OP Stack contracts to chain contracts type. 

### Detailed summary
- Added OP Stack contracts to chain contracts type in `src/types/chain.ts`
- Added contracts for L1 OP Stack and L2 OP Stack
- Added addresses for L2 OP Stack contracts

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Review by tmm — COMMENTED** [review]
If we do something like this, might be better if this is a part of `defineChain`'s `config` parameter and inferred onto `Chain` since not all chains would have these contracts.

File: src/types/chain.ts [hunk 1]
```diff
@@ -37,6 +37,41 @@ export type ChainConstants = {
...
+    // L2 OP Stack Contracts
+    // note these are redeploys and always have the same address
+    l2CrossDomainMessenger?: Address
+    l2StandardBridge?: Address
+    optimismMintableERC20Factory?: Address
+    gasPriceOracle?: Address
+    l1Block?: Address
+    l2ToL1MessagePasser?: Address
+    l2Erc721Bridge?: Address
+    optimismMintableErc721Factory?: Address
```
Comment (roninjin10) [inline src/types/chain.ts:74]:
```suggestion
    // L1 OP Stack Contracts
    optimismL1CrossDomainMessenger?: {
      [chainId: number]: unknown
    }
    optimismL1Erc721Bridge?: {
      [chainId: number]: unknown
    }
    optimismL1StandardBridge?: {
      [chainId: number]: unknown
    }
    optimismL2OutputOracle?: {
      [chainId: number]: unknown
    }
    optimismMintableErc20Factory?: {
      [chainId: number]: unknown
    }
    optimismPortal?: {
      [chainId: number]: unknown
    }
    optimismSystemConfig?: {
      [chainId: number]: unknown
    }
    optimismSystemDictator?: {
      [chainId: number]: unknown
    }
    // L2 OP Stack Contracts
    // note these are redeploys and always have the same address
    optimismL2CrossDomainMessenger?: Address
    optimismL2StandardBridge?: Address
    optimismMintableERC20Factory?: Address
    optimismGasPriceOracle?: Address
    optimismL1Block?: Address
    optimismL2ToL1MessagePasser?: Address
    optimismL2Erc721Bridge?: Address
    optimismMintableErc721Factory?: Address
```
I think we should prefix all optimism addresses with optimism or OP.
Should the address types be `Address | undefined`?

**Review by roninjin10 — CHANGES_REQUESTED** [review]
this is amazing! Long overdue

**Comment by wilsoncusack** [comment]
> If we do something like this, might be better if this is a part of `defineChain`'s `config` parameter and inferred onto `Chain` since not all chains would have these contracts.

That makes sense, thanks. Doing that well may be beyond my TS abilities :) Will also wait for @jxom to chime in as I am porting over some work he had done.

**Comment by tmm** [comment]
@wilsoncusack no worries. @jxom and i can discuss and see what the best approach is.

File: src/types/chain.ts (hunk 1 above)
Comment (wilsoncusack) [reply src/types/chain.ts:74]: yes good call

**Comment by jxom** [comment]
Closing for now. Decided it is best for these to belong in `op-viem`.

**Pull request closed without merging**